### PR TITLE
Ssc 379 ancillary losses

### PR DESCRIPTION
**Pull request by brtietz**
Solves #379. Connects battery's loss model to powerflow via a line in dispatch_t::runDispatch. Losses are covered by:

- PV or the grid when charging or idle
- The battery when discharging

Losses are applied on the AC side for AC connected batteries, and on the DC side for DC connected batteries. AC batteries will not cover the losses from PV unless they can charge from PV, but DC batteries will always cover the charging/idle losses from PV if PV is being exported to load/the grid.

Automated and custom dispatch have new code that accounts for losses. Front of meter automatic, behind the meter automatic, and manual dispatch all now have tests with losses.

### CODE REVIEW

File: shared/lib_battery_powerflow.cpp [hunk 1]
```diff
@@ -154,358 +154,401 @@ double BatteryPower::adjustForDCEfficiencies(double power) {
...
-        if (!m_BatteryPower->canGridCharge && P_grid_to_batt_dc > tolerance){
+    // Quantities are AC in KW unless otherwise specified
+    double P_load_ac = m_BatteryPower->powerLoad;
+    double P_battery_ac, P_pv_ac, P_gen_ac, P_pv_to_batt_ac, P_grid_to_batt_ac, P_batt_to_load_ac, P_grid_to_load_ac, P_pv_to_load_ac, P_pv_to_grid_ac, P_batt_to_grid_ac, P_grid_ac, P_conversion_loss_ac;
+    P_battery_ac = P_pv_ac = P_pv_to_batt_ac = P_grid_to_batt_ac = P_batt_to_load_ac = P_grid_to_load_ac = P_pv_to_load_ac = P_pv_to_grid_ac = P_batt_to_grid_ac = P_gen_ac = P_grid_ac = P_conversion_loss_ac = 0;
+
+    // Quantities are DC in KW unless otherwise specified
+    double P_pv_to_batt_dc, P_grid_to_batt_dc, P_pv_to_inverter_dc;
+    P_pv_to_batt_dc = P_grid_to_batt_dc = P_pv_to_inverter_dc = 0;
+    double P_system_loss_dc = m_BatteryPower->powerSystemLoss; // Units of power sytem loss match battery connection type, see help for details
```
Comment (dguittet) [inline shared/lib_battery_powerflow.cpp:370]: Okay, I see this now-- can you add this to the var_info tables for the compute modules?

File: shared/lib_battery_dispatch_automatic_fom.cpp [hunk 2]
```diff
@@ -291,12 +291,15 @@ void dispatch_automatic_front_of_meter_t::update_dispatch(size_t year, size_t ho
...
         if (m_batteryPower->connectionMode == AC_CONNECTED){
-            if (m_batteryPower->powerBatteryTarget < 0)
-                m_batteryPower->powerBatteryTarget *= m_batteryPower->singlePointEfficiencyDCToAC;
-            else
-                m_batteryPower->powerBatteryTarget /= m_batteryPower->singlePointEfficiencyDCToAC;
+            m_batteryPower->powerBatteryTarget = m_batteryPower->adjustForACEfficiencies(m_batteryPower->powerBatteryTarget, discharge_loss);
+        }
+        else if (m_batteryPower->powerBatteryTarget > 0) {
+            // Adjust for DC discharge losses
+            m_batteryPower->powerBatteryTarget += discharge_loss;
```
Comment (dguittet) [inline shared/lib_battery_dispatch_automatic_fom.cpp:300]: Are these conversion adjustments are only done for the custom dispatch or are they handled independently by the `if` clause just above this?

File: test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp [hunk 3]
```diff
@@ -522,3 +520,103 @@ TEST_F(AutoBTMTest_lib_battery_dispatch, TestCommercialPeakForecasting) {
...
+        if (h > 6 && h < 18) {
+            batteryPower->powerPV = 100; // Match the predicted PV
+        }
+        else if (h > 18) {
+            batteryPower->powerLoad = 40; // Match the predicted load
+        }
+        dispatchAutoBTM->dispatch(0, h, 0);
+        EXPECT_NEAR(batteryPower->powerBatteryDC, expectedPower[h], 0.5) << " error in expected at hour " << h;
+    }
+}
```
Comment (dguittet) [inline test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp:622]: Do you know if we have a Custom dispatch test anywhere for BTM? Just wanted to make sure that the target is met as it is in the FOM case.

**Review by dguittet — CHANGES_REQUESTED** [review]
Really great job on a confusing issue. Thanks for all the tests. I think the logic is sound, though I had some questions about dispatching in the comments. 

Please also update the var_info table to specify that the number of entries acceptable for the loss schedule is 1 or same as weather file.

File: test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp [hunk 4]
```diff
@@ -184,9 +184,7 @@ TEST_F(AutoBTMTest_lib_battery_dispatch, DispatchAutoBTMPVChargeAndDischargeSubh
                                          50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00,
                                          50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00,
                                          50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00,
-                                         50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 50.00, 52.26, 54.52, 56.77,
-                                         59.03, 61.27, 63.52, 65.76, 68.01, 70.25, 72.48, 74.72, 76.95, 79.18, 81.41,
-                                         83.64, 85.87, 88.10, 90.32, 92.55, 94.77, 95.00};
```
[the remark this answers is not in the document]
Comment (brtietz) [reply test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp:189]: Yes, the test only uses the first 96 values, which is actually slightly less that what's in the vector currently.

File: shared/lib_battery_powerflow.cpp [hunk 5]
```diff
@@ -94,51 +94,51 @@ BatteryPower::BatteryPower(const BatteryPower& orig) {
...
 }
 
-double BatteryPower::adjustForACEfficiencies(double power) {
+double BatteryPower::adjustForACEfficiencies(double power, double loss) {
     if (power > 0) {
-        return power / singlePointEfficiencyDCToAC;
+        return (power + loss) / singlePointEfficiencyDCToAC;
     }
     else {
         return power * singlePointEfficiencyACToDC;
```
[the remark this answers is not in the document]
Comment (brtietz) [reply shared/lib_battery_powerflow.cpp:134]: Charging and idle losses are covered by PV or the grid, so they won't change the battery power, unless there isn't enough PV to meet charging+losses and grid charging is disallowed.

File: shared/lib_battery_dispatch_automatic_btm.cpp [hunk 6]
```diff
@@ -804,15 +804,18 @@ void dispatch_automatic_behind_the_meter_t::check_power_restrictions(double& pow
     power = desiredCurrent * _Battery->V() * util::watt_to_kilowatt;
 }
 
-void dispatch_automatic_behind_the_meter_t::set_battery_power(FILE *p, const bool debug)
+void dispatch_automatic_behind_the_meter_t::set_battery_power(size_t idx, FILE *p, const bool debug)
 {
 	for (size_t i = 0; i != _P_target_use.size(); i++) {
+
+        double discharging_loss = _Battery->getDischargeLoss(idx + i, _dt_hour); // Units are kWac for AC connected batteries, and kWdc for DC connected
+
```
[the remark this answers is not in the document]
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_btm.cpp:812]: See comment on lib_battery_powerflow - the battery is only covering discharging losses, so that's the only thing to plan for at this point in dispatch.

File: shared/lib_battery_dispatch_automatic_fom.cpp (hunk 2 above)
Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_fom.cpp:300]: Only custom dispatch right now. The charging code doesn't change, and discharging is always at `_inverter_paco` so there's no additional bandwidth for adding losses.

Comment (brtietz) [reply shared/lib_battery_dispatch_automatic_fom.cpp:300]: Well shoot, now that I've taken a second look at the DC connected losses, those do need to be adjusted, I'll get to that with the other changes.

File: test/shared_test/lib_battery_powerflow_test.cpp [hunk 7]
```diff
@@ -299,10 +462,73 @@ TEST_F(BatteryPowerFlowTest_lib_battery_powerflow, AC_GridCharging_ExcessLoad)
...
+    EXPECT_NEAR(m_batteryPower->powerPVToGrid, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerGridToBattery, 20.83, error);
+    EXPECT_NEAR(m_batteryPower->powerGridToLoad, 40, error);
+    EXPECT_NEAR(m_batteryPower->powerPVToGrid, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerBatteryToLoad, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerBatteryToGrid, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerConversionLoss, 0.83, error);
+    EXPECT_NEAR(m_batteryPower->powerSystemLoss, 1.0, error);
+
+    gen = m_batteryPower->powerPV + m_batteryPower->powerBatteryAC - m_batteryPower->powerSystemLoss;
```
[the remark this answers is not in the document]
Comment (brtietz) [reply test/shared_test/lib_battery_powerflow_test.cpp:509]: That is correct.

File: shared/lib_battery.cpp [hunk 8]
```diff
@@ -216,18 +216,57 @@ void losses_t::run_losses(size_t lifetimeIndex, double dtHour, double charge_ope
...
+    auto hourOfYear = (size_t)std::floor(indexYearOne * dtHour);
+    size_t monthIndex = util::month_of((double)(hourOfYear)) - 1;
+
+    if (params->loss_choice == losses_params::MONTHLY) {
+        return state->loss_kw = params->monthly_idle_loss[monthIndex];
+    }
+    else if (params->loss_choice == losses_params::SCHEDULE) {
+        return params->schedule_loss[lifetimeIndex % params->schedule_loss.size()];
+    }
+}
```
[the remark this answers is not in the document]
Comment (brtietz) [reply shared/lib_battery.cpp:269]: I think when I wrote this I wasn't sure whether the code calling these would know the power. That turned out not to be true, so your changes for encapsulation & reducing duplicated code sound great!

File: shared/lib_battery_dispatch.cpp [hunk 9]
```diff
@@ -30,404 +30,405 @@ OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+    size_t lifetimeIndex = util::lifetimeIndex(year, hour_of_year, step, static_cast<size_t>(1 / _dt_hour));
 
-	do {
+    do {
 
-		// Run Battery Model to update charge based on charge/discharge
-		m_batteryPower->powerBatteryDC = _Battery->run(lifetimeIndex, I);
+        // Run Battery Model to update charge based on charge/discharge
+        m_batteryPower->powerBatteryDC = _Battery->run(lifetimeIndex, I);
+        m_batteryPower->powerSystemLoss = _Battery->get_state().losses->loss_kw;
```
[the remark this answers is not in the document]
Comment (brtietz) [reply shared/lib_battery_dispatch.cpp:411]: battery_t does not provide access to losses_t directly, hence going through the state. Sounds like I should add a new function to battery_t that queries losses_t->getLoss() for the caller?

[the remark this answers is not in the document]
Comment (dguittet) [reply shared/lib_battery_dispatch.cpp:411]: yup!

File: shared/lib_battery_powerflow.cpp [hunk 10]
```diff
@@ -154,358 +154,401 @@ double BatteryPower::adjustForDCEfficiencies(double power) {
...
+                if (P_available_pv < 0.0) {
+                    P_pv_to_batt_ac = 0.0;
+                }
+                else {
+                    P_pv_to_batt_ac = P_available_pv;
+                }
+            }
+        }
+        // Apply PV to losses if possible, grid (via gen) if not
+        bool pv_handles_losses = P_pv_to_batt_ac > 0 || (P_available_pv > 0.0 && P_system_loss_ac > 0.0);
```
[the remark this answers is not in the document]
Comment (brtietz) [reply shared/lib_battery_powerflow.cpp:262]: Good catch! Simplified to `bool pv_handles_losses = P_pv_to_batt_ac > 0 || P_available_pv > 0.0;`

P_pv_to_batt_ac has three possible values on this line:

- 0
- P_battery_ac
- P_available_pv

File: test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp (hunk 3 above)
Comment (brtietz) [reply test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp:622]: We did not, I added them in the latest commit. Great catch, because I had forgotten to update the custom dispatch code for behind the meter, that's fixed now.

I also noticed we're missing tests for grid power targets, so I added an issue to cover that in the future: https://github.com/NREL/ssc/issues/469

Comment (dguittet) [reply test/shared_test/lib_battery_dispatch_automatic_btm_test.cpp:622]: Thanks

**Review by dguittet — APPROVED** [review]
Thanks for the changes! This looks great! Just one more documentation related thing, could you add a comment in battery_t and/or powerflow to explain which losses are covered by which.
